homeassistant/components/synology_dsm/sensor.py

```diff
@@ -0,0 +1,258 @@
+"""Support for Synology DSM Sensors."""
+from datetime import timedelta
+from typing import Dict
+
+from synology_dsm import SynologyDSM
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import (
+    ATTR_ATTRIBUTION,
+    CONF_API_VERSION,
+    CONF_DISKS,
+    CONF_HOST,
+    CONF_NAME,
+    CONF_PASSWORD,
+    CONF_PORT,
+    CONF_SSL,
+    CONF_USERNAME,
+    DATA_MEGABYTES,
+    DATA_RATE_KILOBYTES_PER_SECOND,
+    TEMP_CELSIUS,
+)
+from homeassistant.helpers.dispatcher import async_dispatcher_connect, dispatcher_send
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.event import async_track_time_interval
+from homeassistant.helpers.typing import HomeAssistantType
+
+from .const import (
+    CONF_VOLUMES,
+    DEFAULT_DSM_VERSION,
+    DOMAIN,
+    SERVICE_UPDATE,
+    STORAGE_DISK_SENSORS,
+    STORAGE_VOL_SENSORS,
+    TEMP_SENSORS_KEYS,
+    UTILISATION_SENSORS,
+)
+
+ATTRIBUTION = "Data provided by Synology"
+
+SCAN_INTERVAL = timedelta(minutes=15)
+
+
+async def async_setup_entry(
+    hass: HomeAssistantType, entry: ConfigEntry, async_add_entities
+) -> None:
+    """Set up the Synology NAS Sensor."""
+    name = entry.data[CONF_NAME]
+    host = entry.data[CONF_HOST]
+    port = entry.data[CONF_PORT]
+    username = entry.data[CONF_USERNAME]
+    password = entry.data[CONF_PASSWORD]
+    unit = hass.config.units.temperature_unit
+    use_ssl = entry.data[CONF_SSL]
+    api_version = entry.data.get(CONF_API_VERSION, DEFAULT_DSM_VERSION)
+
+    api = SynoApi(hass, host, port, username, password, unit, use_ssl, api_version)
+
+    await hass.async_add_executor_job(api.update)
+
+    sensors = [
+        SynoNasUtilSensor(api, name, sensor_type, UTILISATION_SENSORS[sensor_type])
+        for sensor_type in UTILISATION_SENSORS
+    ]
+
+    # Handle all volumes
+    if api.storage.volumes_ids:
+        for volume in entry.data.get(CONF_VOLUMES, api.storage.volumes_ids):
+            sensors += [
+                SynoNasStorageSensor(
+                    api, name, sensor_type, STORAGE_VOL_SENSORS[sensor_type], volume
+                )
+                for sensor_type in STORAGE_VOL_SENSORS
+            ]
+
+    # Handle all disks
+    if api.storage.disks_ids:
+        for disk in entry.data.get(CONF_DISKS, api.storage.disks_ids):
+            sensors += [
+                SynoNasStorageSensor(
+                    api, name, sensor_type, STORAGE_DISK_SENSORS[sensor_type], disk
+                )
+                for sensor_type in STORAGE_DISK_SENSORS
+            ]
+
+    async_track_time_interval(hass, api.update, SCAN_INTERVAL)
```

Comment:
Need to store and cancel this when the config entry gets unloaded

Reply:
Or use `DataUpdateCoordinator`
https://developers.home-assistant.io/docs/integration_fetching_data/